// babel-plugin-transform-es2015-modules-ui5/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

exports.default = function (_ref) {
  var t = _ref.types;

  function isValidRequireCall(path) {
    if (!path.isCallExpression()) return false;
    if (!path.get("callee").isIdentifier({ name: "require" })) return false;
    if (path.scope.getBinding("require")) return false;

    var args = path.get("arguments");
    if (args.length !== 1) return false;

    var arg = args[0];
    if (!arg.isStringLiteral()) return false;

    return true;
  }

  var ui5Visitor = {
    MemberExpression: function MemberExpression(path) {
      // check whether module has default and/or named exports
      var node = path.node;
      if (t.isMemberExpression(node.object) && node.object.object.name === 'sap' && node.object.property.name === 'ui' && node.property.name === 'define') {
        this.isAlreadyUI5Module = true;
      }
      if (node.object.name === 'exports' && node.property.name === 'default') {
        this.hasDefaultExport = true;
      }
      if (node.object.name === 'exports' && node.property.name !== 'default') {
        this.anyNamedExport = node.property.name;
      }
    },
    CallExpression: function CallExpression(path) {
      if (!isValidRequireCall(path)) return;
      this.bareSources.push(path.node.arguments[0]);
      path.remove();
    },
    VariableDeclarator: function VariableDeclarator(path) {
      var id = path.get("id");
      if (!id.isIdentifier()) return;

      var init = path.get("init");
      if (!isValidRequireCall(init)) return;

      var source = init.node.arguments[0];
      this.sourceNames[source.value] = true;
      this.sources.push([id.node, source]);

      path.remove();
    }
  };

  return {
    inherits: require("babel-plugin-transform-es2015-modules-commonjs"),

    pre: function pre() {
      // source strings
      this.sources = [];
      this.sourceNames = Object.create(null);

      // bare sources
      this.bareSources = [];

      //
      this.hasDefaultExport = false;
      this.anyNamedExport = undefined;
    },


    visitor: {
      Program: {
        exit: function exit(path) {
          var _this = this;

          if (this.ran) return;
          this.ran = true;

          path.traverse(ui5Visitor, this);

          if (this.isAlreadyUI5Module) {
            return;
          }

          if (this.hasDefaultExport && this.anyNamedExport) {
            throw path.buildCodeFrameError("\n          Your module specifies a default export and a named export: " + this.anyNamedExport + ". \n          When targeting the UI5 module system you can only have a default export ('export default ...') OR multiple named exports ('export const foo = \"bar\"') per module, bot not both.");
          }

          var params = this.sources.map(function (source) {
            return source[0];
          });
          var sources = this.sources.map(function (source) {
            return source[1];
          });

          sources = sources.concat(this.bareSources.filter(function (str) {
            return !_this.sourceNames[str.value];
          }));

          var moduleName = this.getModuleName();
          if (moduleName) moduleName = t.stringLiteral(moduleName);

          if (this.hasDefaultExport || this.anyNamedExport) {
            var exportsIdentifier = t.identifier('exports');

            var exportInitializer = t.variableDeclaration('var', [t.variableDeclarator(exportsIdentifier, t.objectExpression([]))]);
            path.unshiftContainer('body', exportInitializer);

            if (this.hasDefaultExport) {
              path.pushContainer('body', t.returnStatement(t.memberExpression(exportsIdentifier, t.identifier('default'))));
            } else {
              path.pushContainer('body', t.returnStatement(exportsIdentifier));
            }
          }

          var node = path.node;

          var factory = buildFactory({
            PARAMS: params,
            BODY: node.body
          });
          factory.expression.body.directives = node.directives;
          node.directives = [];

          node.body = [buildDefine({
            MODULE_NAME: moduleName,
            SOURCES: sources,
            FACTORY: factory
          })];
        }
      }
    }
  };
};

var _babelTemplate = require("babel-template");

var _babelTemplate2 = _interopRequireDefault(_babelTemplate);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var buildDefine = (0, _babelTemplate2.default)("\n  sap.ui.define(MODULE_NAME, [SOURCES], FACTORY);\n");

var buildFactory = (0, _babelTemplate2.default)("\n  (function (PARAMS) {\n    BODY;\n  })\n");
